### DISCUSSION TITLE
Question regarding internal workings of code actions.

### DISCUSSION THREAD

**Discussion by asmodeus812** [mfussenegger/nvim-jdtls#407] [General]
Hi, Hope this question makes sense. I have recently started looking into how the code actions in nvim-jdtls are implemented and other plugins which integrate with other language servers (such as tsserver https://github.com/jose-elias-alvarez/typescript.nvim ) and find it very weird how code actions are very much handled on per language server basis. They seem to not follow any convention. For example 

For jdtls - i noticed that we have a mapping inside vim.lsp.commands which basically maps a command action string to a lua function which in terms calls the language server anyway, with another action (e.g java.action.hashCodeEqualsPrompt java/checkHashCodeEqualsStatus). So when we call the lsp.buf.code_action method, nvim's internal lua impl fetches the code actions list from jdtls, which provide a list of code actions with the keys/ids which are the java.action.[some-action-name] format, then we map that through vim.lsp.commands to custom lua functions which call other actions for example java/[action-name]. 

For tsserver - the server does not seem to provide any list of actions when you call lsp.buf.code_action method (over file scope), you have to make your own actions (see above the typescript plugin). That is depending on the context, if you have something like unused variable action and you are on that line, you can call that and surely remove it directly by querying the server. But in the same breath organizeImports which should be considered global in scope is not available, and presumably if it was it should be mapped with lsp.commands to be consistent, maybe i have no clue actually it's all over the place for each language server ? See paragraph below.

I am kind of confused, as i thought there is a one to one mapping of what lsp.buf.code_action provides as a list of actions supported for the current context and what the server exposes, rather it seems those are two different things, and there is this weird mapping with vim.lsp.commands, which glues together some of the **_actions mostly with global/file scope / context it seems_** to their behaviour with this artificial mapping, like generating constructors getters/setters etc. https://git.packetlostandfound.us/chiefnoah/neovim/src/commit/433bda405e1fa82db23660a73f275cac5ecca95e/runtime/lua/vim/lsp/handlers.lua#L146

My issue with that is it seems very random and undocumented, code actions seems to be a mashup of both contextual and non contextual actions, for some servers global/file scoped actions are shown (fetched/provided by the server), but not for others. For a **protocol** most implementations seem to be doing whatever the hell they want.  There is no one consistent reliable way to make this work across all servers. We have to use these glue plugins to solve a problem which should have been solved by lsp itself already.

**Comment by mfussenegger** [comment]
> For a protocol most implementations seem to be doing whatever the hell they want. There is no one consistent reliable way to make this work across all servers. We have to use these glue plugins to solve a problem which should have been solved by lsp itself already.


The protocol intentionally left the option for client side extensions. Any compliant server should work with a compliant client without any additions, but having an extension can enable additional functionality.


The important piece is the [Command](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#command) definition in the specification. It can be part of a code action and the spec says this:

> The recommended way to handle commands is to implement their execution on the server side if the client and server provides the corresponding capabilities. Alternatively the tool extension code could handle the command. The protocol currently doesn’t specify a set of well-known commands.


`vim.lsp.commands` is a registry for client side commands. If you use `vim.lsp.buf.code_action` and select a code-action that has a `Command`, neovim will first check if there's an entry in `vim.lsp.commands` and use it, otherwise send the command to the server using [executeCommand](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_executeCommand)


The advantage of client side commands is that they can allow additional functionality. E.g.  `java.action.hashCodeEqualsPrompt` can prompt for the properties that should be included in the generated method. That wouldn't be possible with a server-side-only version.

**Reply by asmodeus812** [reply]
I am fine with the client having the custom handler like what **vim.lsp.commands** does to allow for user interaction before the actual action is executed, what i am not fine with is the complete mess of what is returned / considered an action by the server when you request them (jdtls being a prime example here, since most other servers do not seem to return unimplemented/placeholder actions). See below

> `vim.lsp.commands` is a registry for client side commands. If you use `vim.lsp.buf.code_action` and select a code-action that has a `Command`, neovim will first check if there's an entry in `vim.lsp.commands` and use it, otherwise send the command to the server using [executeCommand](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_executeCommand)

Well yes, that much was obvious from the reference to the source code i posted above, which means that the server returns a list to the client of both **non-implemented** that it knows nothing about, and implemented commands that it can handle. Basically those that can not be executed by the server and are still returned are nothing more than **placeholders**, you can not call the server, but you have to "**know**" they have no handler on the server side (somehow?), in the same breath, other servers **MIGHT** not return any placeholder actions at all (tsserver), so where is the rule here ? . I would have preferred for those to not be returned by **ANY** server at all when you do a call to **buf.code_action**, and the plugin or whatever integrates with the server (in this case jdtls, add them/register them dynamically similarly to what null ls can do, the actions for hashCodeEqualsPrompt for example, the ones that need user interaction before you can call the **ACTUAL** server action with args ). Then we will only interact and enhance only the  actual commands that the server exposes and can actually execute/handle.
